doc/man/croma-message.1
.\" Man page for croma-message.
.TH CROMA-MESSAGE 1 "8 February 2014" "CAFE Desktop Environment"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH "NAME"
croma\-message \- A command to send messages to Croma.
.SH "SYNOPSIS"
.B croma\-message [OPTIONS]
.SH "DESCRIPTION"
The \fBcroma\-message\fP command is used to perform some simple window manager tasks while Croma is running.

.SH "OPTIONS"
.TP
\fBrestart\fR
Restart a currently running instance of \fBcroma\fP.
.TP
\fBreload\-theme\fR
Reload the current window border theme specified in the dconf database. (/org/cafe/croma/general/theme)
.TP
\fBenable\-keybindings\fR
Enable all keybindings specified in the dconf database. (/org/cafe/croma/)
.TP
\fBdisable\-keybindings\fR
Disable all keybindings specified in the dconf database. (/org/cafe/croma/)
.TP
\fBtoggle\-verbose\fR
Enable or Disable debugging messages for \fBcroma\fR.

.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at: 
http://github.com/cafe-desktop/croma/issues
.SH "AUTHORS"
.SS This Man Page has been updated/re-written for the CAFE Desktop Environment by:
Adam Erdman <hekel@archlinux.info> (2014)
.SH "HISTORY"
.SS Portions of this manpage were originally written for the Debian GNU/Linux system by:
Akira TAGOH <tagoh@debian.org> (2002)
.SH "SEE ALSO"
.SS Further information may also be available at: http://wiki.cafe-desktop.org/docs
.P
.BR croma (1),
.BR croma-theme-viewer (1),
.BR croma-window-demo (1)


NEWS
### croma 1.25.0

  * initial commit: marco 1.24.1 fork
  * update FUNDING.yml
  * draw-workspace: Fix memory leak
  * ci: Enable Github Actions
  * display: Safeguard against calling gdk_x11_window_get_xid with a NULL GdkWindow
  * rename marco -> croma
  * rename mate -> cafe
  * port from gtk to ctk
  * themes: rename with ctk
  * prefs: rename 'org.cafe->mate.peripherals-mouse'
    it will be reverted when the mate-control-center fork is done
  * build: avoid deprecated AC_HEADER_STDC, AC_HELP_STRING, AM_PROG_LIBTOOL
  * src,makefile: Fix warnings with %
  * po,Makevars: suppress progress indicators
  * window-props: avoid deprecated 'g_memdup'
  * gschema: set 'ClassicOk' as default theme
  * update AUTHORS
  * reduce the scope of some variables
  * README: remove travis build status
  * Fix [-Wmissing-field-initializers] warnings
  * update translations


src/tools/croma-message.c
/* Croma send-magic-messages app */

/*
 * Copyright (C) 2002 Havoc Pennington
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include <config.h>
#include <ctk/ctk.h>
#include <cdk/cdkx.h>
#include <stdlib.h>
#include <string.h>

#include <libintl.h>
#define _(x) dgettext (GETTEXT_PACKAGE, x)
#define N_(x) x


static void
send_restart (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = CDK_DISPLAY_XDISPLAY (cdk_display_get_default ());
  xev.xclient.window = cdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
                                          "_CROMA_RESTART_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
              cdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()));
  XSync (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()), False);
}

static void
send_reload_theme (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = CDK_DISPLAY_XDISPLAY (cdk_display_get_default ());
  xev.xclient.window = cdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
                                          "_CROMA_RELOAD_THEME_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
              cdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()));
  XSync (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()), False);
}

static void
send_set_keybindings (gboolean enabled)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = CDK_DISPLAY_XDISPLAY (cdk_display_get_default ());
  xev.xclient.window = cdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
                                          "_CROMA_SET_KEYBINDINGS_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = enabled;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
              cdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()));
  XSync (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()), False);
}

#ifdef WITH_VERBOSE_MODE
static void
send_toggle_verbose (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = CDK_DISPLAY_XDISPLAY (cdk_display_get_default ());
  xev.xclient.window = cdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
                                          "_CROMA_TOGGLE_VERBOSE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()),
              cdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()));
  XSync (CDK_DISPLAY_XDISPLAY (cdk_display_get_default ()), False);
}
#endif

static void
usage (void)
{
  g_printerr (_("Usage: %s\n"),
              "croma-message (restart|reload-theme|enable-keybindings|disable-keybindings|toggle-verbose)");
  exit (1);
}

int
main (int argc, char **argv)
{
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");

  ctk_init (&argc, &argv);

  if (argc < 2)
    usage ();

  if (strcmp (argv[1], "restart") == 0)
    send_restart ();
  else if (strcmp (argv[1], "reload-theme") == 0)
    send_reload_theme ();
  else if (strcmp (argv[1], "enable-keybindings") == 0)
    send_set_keybindings (TRUE);
  else if (strcmp (argv[1], "disable-keybindings") == 0)
    send_set_keybindings (FALSE);
  else if (strcmp (argv[1], "toggle-verbose") == 0)
    {
#ifndef WITH_VERBOSE_MODE
      g_printerr (_("Croma was compiled without support for verbose mode\n"));
      return 1;
#else
      send_toggle_verbose ();
#endif
    }
  else
    usage ();

  return 0;
}

